crates/valence_command/src/parser.rs

```diff
@@ -0,0 +1,232 @@
+use std::borrow::Cow;
+use std::ops::Range;
+
+use valence_core::text::Text;
+
+use crate::parsing_ret_err;
+use crate::reader::StrReader;
+
+#[derive(Clone, Debug)]
+pub struct Suggestion<'a> {
+    pub message: Cow<'a, str>,
+    pub tooltip: Option<Text>,
+}
+
+impl<'a> Suggestion<'a> {
+    pub const fn new_str(message: &'a str) -> Self {
+        Self {
+            message: Cow::Borrowed(message),
+            tooltip: None,
+        }
+    }
+}
+
+impl<'a> From<&'a str> for Suggestion<'a> {
+    fn from(value: &'a str) -> Self {
+        Self {
+            message: Cow::Borrowed(value),
+            tooltip: None,
+        }
+    }
+}
+
+impl<'a> From<String> for Suggestion<'a> {
+    fn from(value: String) -> Self {
+        Self {
+            message: Cow::Owned(value),
+            tooltip: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ParsingResult<T, S, E> {
+    pub suggestions: Option<(Range<usize>, S)>,
+    pub result: Result<Option<T>, (Range<usize>, E)>,
+}
+
+impl<T, S, E> ParsingResult<T, S, E> {
+    pub const fn ok() -> Self {
+        Self {
+            suggestions: None,
+            result: Ok(None)
+        }
+    }
+
+    pub fn map_suggestion<S1>(self, func: impl FnOnce(S) -> S1) -> ParsingResult<T, S1, E> {
+        ParsingResult {
+            suggestions: self.suggestions.map(|(pos, s)| (pos, func(s))),
+            result: self.result,
+        }
+    }
+
+    pub fn map_ok<T1>(self, func: impl FnOnce(T) -> T1) -> ParsingResult<T1, S, E> {
+        ParsingResult {
+            suggestions: self.suggestions,
+            result: self.result.map(|v| v.map(func)),
+        }
+    }
+
+    pub fn map_err<E1>(self, func: impl FnOnce(E) -> E1) -> ParsingResult<T, S, E1> {
+        ParsingResult {
+            suggestions: self.suggestions,
+            result: self.result.map_err(|(pos, e)| (pos, func(e))),
+        }
+    }
+
+    pub fn zip<T1>(
+        self,
+        func: impl FnOnce() -> ParsingResult<T1, S, E>,
+    ) -> ParsingResult<(T, T1), S, E> {
+        let res = parsing_ret_err!(self);
+
+        let other = func();
+
+        ParsingResult {
+            suggestions: other.suggestions,
+            // SAFETY: parsing_ret_err! ensures that result is Ok
+            result: other
+                .result
+                .map(|v| match (v, unsafe { res.result.unwrap_unchecked() }) {
+                    (Some(v), Some(o)) => Some((o, v)),
+                    _ => None,
+                }),
+        }
+    }
+}
+
+/// Macro ensures that returned result contains no Error
+#[macro_export]
+macro_rules! parsing_ret_err {
```

Comment:
I strongly dislike declarative macros (especially those in public APIs) unless it's absolutely necessary. Declarative macros should be tools of last resort. I suggest finding a way to avoid these kinds of macros, but this indicates to me that something about `ParsingResult` is more complicated than it should be.